### DISCUSSION TITLE
Foreign key constraint failed on the field: `foreign key`

### DISCUSSION THREAD

**Discussion by kivox** [prisma/prisma#10867] [Q&A] [answered]
Hi!

I'm trying to delete a "project" row but get a warning for Foreign key constraints. This only happened after I added the member model as I was previously connecting users directly to the project model, this is because I'd like to add a permission system and other things on a per user basis.

schema.prisma
```prisma
model User {
  id        Int       @id @unique @default(autoincrement())
  email     String    @unique
  projects  Member[]  @relation("user_projects")
  owner     Project[] @relation("project_owner")
  invites   Invite[]  @relation("invitee")
}

model Project {
  id        Int      @id @unique @default(autoincrement())
  name      String   @default("Unnamed Project")
  members   Member[] @relation("project_members")
  invites   Invite[] @relation("project_invites")
  apps      App[]    @relation("project_apps")
  owner     User     @relation("project_owner", fields: [ownerID], references: [id])
  ownerID   Int
}

model Invite {
  id        Int      @id @unique @default(autoincrement())
  invitee   User     @relation("invitee", fields: [inviteeID], references: [id])
  inviteeID Int
  project   Project  @relation("project_invites", fields: [projectID], references: [id])
  projectID Int
  email     String
}

model Member {
  id        Int      @id @unique @default(autoincrement())
  project   Project  @relation("project_members", fields: [projectID], references: [id])
  projectID Int
  user      User     @relation("user_projects", fields: [userID], references: [id])
  userID    Int
}

model App {
  id        Int      @id @unique @default(autoincrement())
  name      String
  project   Project  @relation("project_apps", fields: [projectID], references: [id])
  projectID Int
}
```

This is how I query it:
```js
await prisma.project.delete({
	where: {
		id: parseInt(req.params.project)
	},
})
```

output:
```
DELETE: /projects/1

PrismaClientKnownRequestError: 
Invalid `prisma.project.delete()` invocation in
/home/src/api/routes/projects/delete.js:6:46

  3 
  4 module.exports = async (req, res) => {
  5     try {
→ 6             const project = await prisma.project.delete(
  Foreign key constraint failed on the field: `foreign key`
    at cb (/home/src/api/node_modules/@prisma/client/runtime/index.js:38683:17)
    at async module.exports (/home/src/api/routes/projects/delete.js:6:19) {
  code: 'P2003',
  clientVersion: '3.6.0',
  meta: { field_name: 'foreign key' }
}
```

**Comment by janpio** [comment, marked as answer]
This might help: https://www.prisma.io/docs/concepts/components/prisma-schema/relations/referential-actions

**Reply by davidyell** [reply]
I implemented this approach, but still ended up here because of the same issue.

(Apologies for the naming, as I am learning Prisma still 🎓 )

I am using SQLite as my data store, [which is supported](https://www.prisma.io/docs/concepts/components/prisma-schema/relations/referential-actions#types-of-referential-actions)

Although my configuration differs slightly, as I am trying to delete a One to Many. My expectation is that deleting a `Decks` would cascade delete all the related `CardsInDecks` records.

```prisma
model CardsInDecks {
  id           Int     @id @default(autoincrement())
  deck_id      Int
  card_id      Int
  quantity     Int     @default(1)
  is_sideboard Boolean @default(false)
  decks        Decks   @relation(fields: [deck_id], references: [id], onDelete: Cascade)
  cards        Cards   @relation(fields: [card_id], references: [id], onDelete: NoAction, onUpdate: NoAction)

  @@index([deck_id, card_id], map: "cards_in_decks_deck_id_IDX")
  @@map("cards_in_decks")
}

model Decks {
  id             Int            @id @default(autoincrement())
  name           String
  player_id      Int
  created        DateTime       @default(now())
  updated        DateTime?      @updatedAt
  player         Players        @relation(fields: [player_id], references: [id], onDelete: NoAction, onUpdate: NoAction)
  cards_in_decks CardsInDecks[]

  @@index([player_id], map: "decks_player_id_IDX")
  @@map("decks")
}
```

I am calling it in the standard way,
```typescript
const result = await prismaClient.decks.delete({
  where: { id: deckId },
});
```

..but I get the error mentioned by OP.
```bash
  152 const remove = async (request: Request, response: Response) => {
  153   const deckId: Decks['id'] = parseInt(request.params.id, 10);
  154
→ 155   const result = await prismaClient.decks.delete(
  Foreign key constraint failed on the field: `foreign key`
```

**Reply by davidyell** [reply]
Apologies, I solved this soon after I wrote this.

The key point I was missing was that I needed to push these schema changes into the database, whereas I was just generating a new client.

So be sure to `yarn prisma db push`

**Comment by Lexpeee** [comment]
I would like to add, I'm currently at a point where I'm still learning how to implement Prisma to my nestjs server project. Here is my schema: 

```prisma

model posts {
  id          String    @id @default(uuid()) @db.Uuid
  userId      String    @db.Uuid
  dateCreated DateTime  @default(now())
  dateRemoved DateTime?
  content     String
  embed       Json?
  attachments Json?
  likes       Int
  user        users     @relation(fields: [userId], references: [id])
}

model users {
  id          String    @id @default(uuid()) @db.Uuid
  username    String
  password    String
  status      String
  dateCreated DateTime  @default(now())
  dateRemoved DateTime?
  posts       posts[]
}
``` 

And the error I got: 
```
const newPost = await this.prisma.posts.create(
Foreign key constraint failed on the field: `posts_userId_fkey (index)`
```
Pardon me for the lack of naming conventions and stuff. lol
I tried adding a new data to the database, this error occurs when the userID i put in doesn't exist in the user database. Just wanting to share my experience for the upcoming learniers. :)

**Reply by Abdul-Manan-456** [reply]
i have faced the same issue, but now solve. thanks..............

**Reply by orlando1807** [reply]
const { raffleId, number, status } = req.body;
const existingRaffle = await prisma.raffles.findUnique({
            where: { id: raffleId },
        });

        if (!existingRaffle) {
            return res.status(404).json({ error: 'Raffle not found' });
        }
        const itemInstances = await prisma.tickets.create({
            data: {
                number,
                status,
                raffle_id:raffleId,
            }
        })
      I always get the same error
      my schema is: 
      model raffles {
  id            String          @id @default(uuid()) @db.Uuid
  name          String?         @db.VarChar
  date          DateTime?       @db.Date
  place         String?         @db.VarChar
  link          String?         @db.VarChar
  price_ticket  Int?
  imagen        String?         @db.VarChar
  quantity      Int?
  created_at    DateTime?       @default(now()) @db.Timestamp(6)
  audit         audit[]
  invoices      invoices[]
  raffle_awards raffle_awards[]
  tickets       tickets[]
}
and
model tickets {
  id              String          @id @default(uuid()) @db.Uuid
  number          Int?
  raffle_id       String          @db.Uuid
  user_id         String?         @db.Uuid
  status          Int?
  date_purchase   DateTime?       @db.Date
  pay_method      String?         @db.Uuid
  picture_id      String?         @db.Uuid
  created_at      DateTime?       @default(now()) @db.Timestamp(6)
  invoice_tickets invoice_tickets @relation(fields: [id], references: [id], onDelete: NoAction, onUpdate: NoAction)
  raffle          raffles          @relation(fields: [raffle_id], references: [id]) 
  users           users?          @relation(fields: [user_id], references: [id], onDelete: NoAction, onUpdate: NoAction)
  payment_method  payment_method? @relation(fields: [pay_method], references: [id], onDelete: NoAction, onUpdate: NoAction)
}

help me please

**Reply by adambak033** [reply]
Thank you very much ❤️

**Reply by Moe03** [reply]
Hey I also faced this issue for quite sometime **make sure that the parent of the child does exist** so for the above Users, Posts example, a user must exist first and be created if its not there that error will appear, also if you generated an id (for some reason) for the parent that doesn't exist it will still throw that error which was what i was doing

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
